src/layout/layout.rs

```diff
@@ -436,333 +443,474 @@ impl Layout {
-    ///     .split(Rect::new(2, 2, 10, 10));
-    /// assert_eq!(layout[..], [Rect::new(2, 2, 10, 5), Rect::new(2, 7, 10, 5)]);
-    ///
-    /// let layout = Layout::default()
-    ///     .direction(Direction::Horizontal)
-    ///     .constraints([Constraint::Ratio(1, 3), Constraint::Ratio(2, 3)])
-    ///     .split(Rect::new(0, 0, 9, 2));
-    /// assert_eq!(layout[..], [Rect::new(0, 0, 3, 2), Rect::new(3, 0, 6, 2)]);
-    /// ```
-    pub fn split(&self, area: Rect) -> Rc<[Rect]> {
-        LAYOUT_CACHE.with(|c| {
-            c.get_or_init(|| {
-                RefCell::new(LruCache::new(
-                    NonZeroUsize::new(Self::DEFAULT_CACHE_SIZE).unwrap(),
-                ))
-            })
-            .borrow_mut()
-            .get_or_insert((area, self.clone()), || {
-                Self::try_split(area, self).expect("failed to split")
-            })
-            .clone()
-        })
-    }
-
-    fn try_split(area: Rect, layout: &Layout) -> Result<Rc<[Rect]>, AddConstraintError> {
-        let mut solver = Solver::new();
-        let inner = area.inner(&layout.margin);
-
-        let (area_start, area_end) = match layout.direction {
+    fn get_start_end_size(&self, area: Rect) -> (f64, f64, f64) {
+        let inner = area.inner(&self.margin);
+        let (area_start, area_end) = match self.direction {
             Direction::Horizontal => (f64::from(inner.x), f64::from(inner.right())),
             Direction::Vertical => (f64::from(inner.y), f64::from(inner.bottom())),
         };
         let area_size = area_end - area_start;
+        (area_start, area_end, area_size)
+    }
 
-        let spacers_added = if layout.spacing == 0 {
-            false
-        } else {
-            matches!(
-                layout.flex,
-                Flex::Stretch | Flex::StretchLast | Flex::Start | Flex::Center | Flex::End
-            )
-        };
-        let constraints = if spacers_added {
-            Itertools::intersperse(
-                layout.constraints.iter().cloned(),
-                Constraint::Fixed(layout.spacing),
-            )
-            .collect_vec()
-        } else {
-            layout.constraints.clone()
-        };
-
+    /// Builds a vector of [`Element`]s corresponding to user provided constraints.
+    ///
+    /// ```plain
+    /// <----------------------------80 px (spacer: 10 px)----------------------------->
+    /// ┌──────────────────┐┌        ┐┌──────────────────┐┌        ┐┌──────────────────┐
+    /// │     Fixed(20)    │          │      Min(20)     │          │      Max(20)     │
+    /// └──────────────────┘└        ┘└──────────────────┘└        ┘└──────────────────┘
+    /// ^^^^^^^^^^^^^^^^^^^^          ^^^^^^^^^^^^^^^^^^^^          ^^^^^^^^^^^^^^^^^^^^
+    ///          │                              │                             │
+    ///          └──────────────────────────────┴───────────────────────────Segments
+    /// ```
```

Comment:
This picture is is a lot of lines just to explain the concept of what a segment is, but doesn't provide any information about the result or actions that this method does.

I find it a bit confusing given the name "build" that this actually both creates and also adds the elements to the solver and then returns them. The interaction between layout, solver, and each element is a bit difficult to visualize here.

Working through this method and the Element::constrain call, I think the net effect of this method is to create variables to track the start and end of the element areas, and then to and cassowary constraints that ensure those variables are within the range.

I think this might be a case where overly splitting and documenting makes it more difficult to sense what's going on here.

How about something like the following, which I think would go a long way to making this generally super easy to follow:

```rust
let (start, end, length) = match (self.direction) {  
    Direction::Horizontal => (inner.left().into(), inner.right().into(), inner.width.into()),  
    Direction::Vertical => (inner.top().into(), inner.bottom().into(), inner.height.into()),  
};
let segment_count = layout.constraints.len();
let spacer_count = segment_count.saturating_add(1);
let segments = iter::repeat_with(Element::new).take(segment_count).collect();
let spacers = iter::repeat_with(Element::new).take(spacer_count).collect();

// ensure that segments and spacers are within the area
solver.add_constraints(segment.iter().map(|e| e.to_within_area_constraint(start, end)));
solver.add_constraints(spacers.iter().map(|e| e.to_within_area_constraint(start, end)));

// ensure the segments and spacers are arranged left to right and interleaved
solver.add_constraints(spacers.iter().interleave(segments.iter()).tuple_windows().map((a,b) colinear_constraint(a, b)));
solver.add_constraints(spacers.iter().first().map(|e| e.start_constraint(start)));
solver.add_constraints(spacers.iter().last().map(|e| e.end_constraint(end)));
```

Reply:
I've updated the code to use `Element` methods.